Trouble importing all of the documents

I will have to separate these commits out a little better. I had trouble importing all of the documents. Not all of them came in when running `npm run bootstrap` or even by simply running `npx gulp importAll`. I think the flaky GitHub downloads leads to winding up w/ documents missing — a bit of a hassle. :/ Anyways, I'll get back to this in a short while. Thank you for your patience.

_Originally posted by @DerekNonGeneric in https://github.com/ampproject/amp.dev/pull/3346#issuecomment-644847783_

To reproduce this bug:

1. `nvm install --lts`
2. `npm install`
3. `npm run lint` (be sure it passes)
4. `npm run bootstrap`
5. `npm run lint` (this fails)

Actual:

```
[12:35:57 PM] [Reference checker] › ✖  error     No matching document found for /content/amp-dev/documentation/components/amp-access/0.1/iframe-api/README.md. Needs manual fixing. (First found in /mnt/c/orgs/ampproject/amp.dev/pages/content/amp-dev/documentation/components/reference/amp-subscriptions-v0.1.md)
[12:35:57 PM] [Reference checker] › ✖  error     No matching document found for /content/amp-dev/documentation/components/amp-subscriptions-google/amp-subscriptions-google.md. Needs manual fixing. (First found in /mnt/c/orgs/ampproject/amp.dev/pages/content/amp-dev/documentation/components/reference/amp-subscriptions-v0.1.md)


[12:36:12] 'lintGrow' errored after 33 s
[12:36:12] Error: 2 broken links and 0 wrong anchors found
    at Pumpify.<anonymous> (/mnt/c/orgs/ampproject/amp.dev/platform/lib/tools/growReferenceChecker.js:177:13)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
[12:36:12] 'lintAll' errored after 33 s
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! amp-dev@1.0.5 lint: `npx gulp lintAll`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the amp-dev@1.0.5 lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

Expected (some similar):

```
✔ success All references intact!
```

--- a/README.md
+++ b/README.md
@@ -104,6 +104,8 @@ This command enables the import from GitHub to run flawlessly. The actual import
 $ npm run bootstrap
 ```
 
+**Tip**: Due to bad network conditions or GitHub's API rate-limiting there might be errors during import. Try running the above command with the `-- --queue-imports` flag to prevent them.
+
 You can then start developing in your local environment with the command below. The task will take care of building and copying all files, watching them for changes, and rebuilding them when needed. Beware that changes to the [Express](https://expressjs.com/) backend require the Gulp task to be restarted.
 
 ```sh

--- a/package.json
+++ b/package.json
@@ -80,6 +80,7 @@
     "node-fetch": "2.6.0",
     "node-zopfli-es": "1.0.4",
     "nunjucks": "3.2.1",
+    "p-queue": "6.4.0",
     "randomstring": "1.1.5",
     "rcs-core": "3.3.1",
     "safe-eval": "0.4.1",

--- a/platform/lib/pipeline/gitHubImporter.js
+++ b/platform/lib/pipeline/gitHubImporter.js
@@ -14,6 +14,7 @@
  * limitations under the License.
  */
 
+const config = require('../config.js');
 const octonode = require('octonode');
 const {default: PQueue} = require('p-queue');
 const {Signale} = require('signale');
@@ -27,7 +28,9 @@ const CLIENT_ID = process.env.AMP_DOC_ID;
 const DEFAULT_ORGANISATION = 'ampproject';
 const DEFAULT_REPOSITORY = `${DEFAULT_ORGANISATION}/amphtml`;
 
-const MAX_CONCURRENT_OPS = 12;
+const MAX_CONCURRENT_OPS = config.options['queue-imports']
+  ? 6
+  : Number.POSITIVE_INFINITY;
 
 const log = new Signale({
   'interactive': false,

--- a/platform/lib/pipeline/gitHubImporter.js
+++ b/platform/lib/pipeline/gitHubImporter.js
@@ -17,6 +17,7 @@
 const fs = require('fs').promises;
 const path = require('path');
 const octonode = require('octonode');
+const {default: PQueue} = require('p-queue');
 const {Signale} = require('signale');
 
 const LOCAL_AMPHTML = process.env.AMP_LOCAL_AMPHTML;
@@ -27,6 +28,8 @@ const CLIENT_ID = process.env.AMP_DOC_ID;
 /* The GitHub organisation where the repositories imported from are located */
 const DEFAULT_ORGANISATION = 'ampproject';
 const DEFAULT_REPOSITORY = `${DEFAULT_ORGANISATION}/amphtml`;
+
+const MAX_CONCURRENT_OPS = 12;
 
 const log = new Signale({
   'interactive': false,
@@ -47,8 +50,8 @@ function checkCredentials() {
 
 class GitHubImporter {
   constructor() {
-    this._log = log;
     checkCredentials();
+    this._queue = new PQueue({concurrency: MAX_CONCURRENT_OPS});
     this._github = octonode.client(
       CLIENT_TOKEN || {
         'id': CLIENT_ID,
@@ -105,30 +108,38 @@ class GitHubImporter {
     master = false,
     local = false
   ) {
+    let request = Promise.resolve();
+
     if (!filePath) {
-      return Promise.reject(new Error('Can not download from undefined path.'));
+      request = Promise.reject(
+        new Error('Can not download from undefined path.')
+      );
+    } else if (local) {
+      // TODO: https://github.com/ampproject/amp.dev/issues/2965
+      request = Promise.reject(
+        new Error('Importing from a local path is currently not implemented.')
+      );
+    } else if (master || repo !== DEFAULT_REPOSITORY) {
+      request = await this._queue.add(() => {
+        log.await(`Downloading ${filePath} from remote master...`);
+        return this._github.repo(repo).contentsAsync(filePath);
+      });
+    } else {
+      const tag = await this._fetchLatestReleaseTag();
+      request = await this._queue.add(() => {
+        log.await(`Downloading ${filePath} from remote [${tag}]...`);
+        return this._github.repo(repo).contentsAsync(filePath, tag);
+      });
     }
 
-    if (local) {
-      this._log.await(`Copying ${filePath} from local file path...`);
-      return fs.readFile(path.join(local, filePath));
-    }
-
-    if (master || repo !== DEFAULT_REPOSITORY) {
-      this._log.await(`Downloading ${filePath} from remote master...`);
-      return this._github.repo(repo).contentsAsync(filePath);
-    }
-
-    const branch = await this._fetchLatestReleaseTag();
-    this._log.await(`Downloading ${filePath} from remote [${branch}]...`);
-    return this._github.repo(repo).contentsAsync(filePath, branch);
+    return request;
   }
 
   _fetchLatestReleaseTag() {
     if (this.latestReleaseTag_) {
       return this.latestReleaseTag_;
     }
-    this._log.await('Fetching latest release tag ...');
+    log.await('Fetching latest release tag ...');
 
     this.latestReleaseTag_ = new Promise((resolve, reject) => {
       this._github.get(
@@ -138,7 +149,7 @@ class GitHubImporter {
           if (body.tag_name) {
             resolve(body.tag_name);
           } else {
-            this._log.fatal(
+            log.fatal(
               'Was not able to retrieve latest AMP release from GitHub.'
             );
             reject(err);
@@ -147,11 +158,11 @@ class GitHubImporter {
       );
     })
       .then((latestReleaseTag) => {
-        this._log.success(`Fetched latest release tag: ${latestReleaseTag}`);
+        log.success(`Fetched latest release tag: ${latestReleaseTag}`);
         return latestReleaseTag;
       })
       .catch((err) => {
-        this._log.fatal(err);
+        log.fatal(err);
         return null;
       });
     return this.latestReleaseTag_;
@@ -160,7 +171,7 @@ class GitHubImporter {
 
 module.exports = {
   CLIENT_TOKEN,
-  CLIENT_SECRET: CLIENT_SECRET,
+  CLIENT_SECRET,
   CLIENT_ID,
   log,
   checkCredentials,

--- a/platform/lib/pipeline/gitHubImporter.js
+++ b/platform/lib/pipeline/gitHubImporter.js
@@ -14,8 +14,6 @@
  * limitations under the License.
  */
 
-const fs = require('fs').promises;
-const path = require('path');
 const octonode = require('octonode');
 const {default: PQueue} = require('p-queue');
 const {Signale} = require('signale');

